Register level gateway

AlessandroA opened the pull request:
This PR introduces the Register Gateway implementation.
- Added the issue #241 to track the runtime OR operation on the gateway function pointer independently of this pull request.

Done so far:

* Remove `uvisor_{read,write}` APIs restrictions (see #90) by using  a `struct` instead of SVCall metadata.
* All operations implemented.
* Added APIs for common bit manipulations.
* All operations tested.

Please concentrate on the last 2 commits.

@meriac @Patater @niklas-arm

Patater commented on api/inc/register_gateway.h:
```diff
@@ -17,105 +17,146 @@
+        }; \
+        \
+        /* Call the actual gateway. */ \
+        uint32_t result = ((uint32_t (*)(void)) ((uint32_t) &register_gateway | 1))(); \
```
So, just casting to the function pointer doesn't set the thumb bit. We have to do this manually as done here. Correct?

AlessandroA replied:
Yes. At the moment I've tried the following:

* Just cast to a fn pointer. Doesn't work (Thumb bit not set).
* Save `gtw_ptr | 1` in a `static const` and use that as fn ptr. Doesn't work (`gtw_ptr` not known until link time, the `| 1` is done at runtime.

Please feel free to suggest other approaches!

Patater commented on api/inc/register_gateway.h:
```diff
@@ -17,105 +17,146 @@
+ * void uvisor_read(src_box, address);
+ * void uvisor_read(src_box, address, operation, mask);
+ *
+ * @warning These APIs currently only support link-time known values. This means
```
What does this warning mean for the `uvisor_read` operation? The following example is for `uvisor_write`. Consider either removing the warning to making the example illustrate what the warning means.

AlessandroA replied:
Updated.

Patater commented on api/inc/register_gateway_exports.h:
```diff
@@ -0,0 +1,59 @@
+ * The 'x's can be freely chosen (they have been chosen randomly here).
+ * This requires the Thumb-2 extensions because otherwise only 16 bits opcodes
+ * are accepted.*/
+#if defined(__thumb__) && defined(__thumb2__)
```
Why do we have to check both __thumb__ and __thumb2__? Why not just check __thumb2__?

AlessandroA replied:
It is to make sure to select the configuration we know at the moment (Thumb with Thumb-2 extension). Currently that is identical to just having `__thumb2__`, but I thought it was better to spell it out.

Patater commented on api/inc/register_gateway.h:
```diff
@@ -17,105 +17,146 @@
-
-#endif /* __CC_ARM or __GNUC__ */
+/** Both `uvisor_read` and `uvisor_write` can have 2 additional input arguments -
+ *  mask and operation. These macros select the right argument from
```
I'd rather not we have optional macro arguments, for code maintainability sake. We need to balance API ease of understanding and use with code maintenance and future refactoring efforts.

I'd prefer named macros like `uvisor_write_mask` that takes a mask and `uvisor_write_op` that takes an op, and `uvisor_write_mask_op` that takes  both. The macros that accept fewer parameters can be implemented in terms of the macros that take more parameters; this maintains the aspect of your current implementation where the default values for parameters easy to identify.

meriac replied:
@AlessandroA I agree with Jaeden here - that API won't be directly exposed to the user anyway. It's much clearer if we move the macro indirection.

AlessandroA replied:
I removed the variadic macros and only left the full ones. On the other hand, I provided new APIs to perform common operations as set, clear, toggle bits.

Patater commented on core/system/src/register_gateway.c:
```diff
@@ -0,0 +1,95 @@
+ * @internal
+ * @warning This function does not verify that the operation, value and mask
+ * make sense together. The code that performs the operation should check for
+ * this and throw an error accordingly.
```
s/throw/return

Patater commented on core/system/src/mpu/vmpu.c:
```diff
@@ -487,3 +482,9 @@ int vmpu_box_namespace_from_id(int box_id, char *box_namespace, size_t length)
 }
+
+/* TODO */
+void vmpu_register_gateway_check(uint32_t address, uint32_t operation, uint32_t mask)
```
Can we delete this function? What would we need it for?

niklas-arm commented on api/inc/register_gateway.h:
```diff
@@ -17,104 +17,146 @@
+ * @internal
+ */
+#define SELECT_MASK(...)       __UVISOR_MACRO_SELECT(_0, ##__VA_ARGS__, , , _MASK , , _DEFAULT_NULL)(__VA_ARGS__)
+#define SELECT_READ_OP(...)    __UVISOR_MACRO_SELECT(_0, ##__VA_ARGS__, , , _OP , , _DEFAULT_READ)(__VA_ARGS__)
```
Note that the `@internal` only applies to `SELECT_MASK(...)`. You can group it (`@{`,`@}`).

AlessandroA replied:
Right! Thanks!

AlessandroA commented:
@meriac @Patater @niklas-arm 

* Addressed comments from first round of review.
* Added implementation of all operations.
* Added APIs for common bit manipulations.

Remaining:

* Writing tests for all operations.

**Note**: A linker script update is required.

Patater commented on core/system/inc/linker.h:
```diff
@@ -19,62 +19,69 @@
 
-        /* pointers to boxes configuration tables */
-    uint32_t *cfgtbl_ptr_start, *cfgtbl_ptr_end;
+        /* Pointers to the secure boxes configuration tables */
```
Do we want comments not aligned with the code still, or should we fix that in this commit, too?

AlessandroA replied:
I removed the indentation, I think it's cleaner! Thanks for spotting it

Patater commented on api/src/uvisor-input.S:
```diff
@@ -37,54 +40,49 @@ uvisor_init:
 
-    /* start and end address of uvisor main bss */
+        /* uVisor own BSS section */
     .long __uvisor_bss_main_start
```
Do we want comments to line up with the code here or not?

Patater commented on api/src/uvisor-input.S:
```diff
@@ -37,54 +40,49 @@ uvisor_init:
     .long __uvisor_bss_end
 
-    /* start and end address of uvisor main bss */
+        /* uVisor own BSS section */
```
Do we want comments to line up with the code here or not?

AlessandroA replied:
I removed the indentation, I think it's cleaner! Thanks for spotting it

Patater commented on api/src/uvisor-input.S:
```diff
@@ -93,14 +91,23 @@ uvisor_config:
+/* uVisor mode of operation
+ * 0: Disabled
+ * 1: Permissive
+ * 2: Enabled */
```
What if we change these values elsewhere? We'd forget to update this comment, probably. For the sake of maintainability, consider removing the values here and reference `enum` names or `#define` names instead, if you want to reference the values that `__uvisor_mode` can be.

AlessandroA replied:
Yeah unfortunately the names would be UVISOR_ENABLED, _DISABLED, _PERMISSIVE and they're not an enum, but just preprocessor definitions. It would be equally outdated in case of change. I'd rather remove the comment altogether, what do you think?

Patater replied:
I've fine having the comment use the `#define` names. Those are part of our API can shouldn't change.

I'm also fine removing the comment, but there must be some value in having the comment as you felt there was a reason to add it.

Patater commented on api/src/uvisor-input.S:
```diff
@@ -93,14 +91,23 @@ uvisor_config:
     .long 0
 
 .section .bss
+/* Pointer to the secure boxes context. It is update at runtime by uVisor to
```
s/update/updated

Patater commented on api/inc/register_gateway.h:
```diff
@@ -17,105 +17,208 @@
+/** Get the offset of a struct member.
+ * @internal
+ */
+#define OFFSETOF(type, member) ((uint32_t) (&(((type *)(0))->member)))
```
Can we use `offsetof` instead of this macro?

AlessandroA replied:
`offsetof` is a GCC builtin and it's not available on ARMCC that I know of. Since this is user code I wanted to make it more general.

Patater replied:
That's unfortunate, for ARMCC.

niklas-arm replied:
Good, I'm going to need this macro for the tier-2 allocator then ;-)

Patater commented on api/inc/register_gateway.h:
```diff
@@ -17,105 +17,208 @@
+        (void) register_gateway_ptr; \
+        \
+        /* Call the actual gateway. */ \
+        uint32_t result = ((uint32_t (*)(void)) ((uint32_t) &register_gateway | 1))(); \
```
I'm fine with this implementation. File an issue for getting rid of this `| 1`. We can then prioritize the optimization appropriately. I wouldn't block this PR based on having the run-time `| 1` here, and it isn't a priority to get rid of it right now.

meriac replied:
Please mention that issue in the comment of the pull request.

Patater commented on api/inc/register_gateway_exports.h:
```diff
@@ -0,0 +1,57 @@
+ * @warning The operation value must be hardcoded. */
+#define UVISOR_RGW_OP_READ          0 /**< value = *address                      */
+#define UVISOR_RGW_OP_READ_AND      1 /**< value = *address & mask               */
+#define UVISOR_RGW_OP_WRITE         4 /**< *address = value                      */
```
Is there some undocumented scheme here behind the numbers? Why no `2` or `3`. Please comment why in the code, if this is on purpose.

AlessandroA replied:
Just a typo. Fixed

Patater commented on api/src/uvisor-input.S:
```diff
@@ -79,6 +79,10 @@ uvisor_config:
     .long __uvisor_cfgtbl_ptr_start
     .long __uvisor_cfgtbl_ptr_end
 
+        /* Pointers to the secure boxes register gateways */
```
Are we sure we don't want to align comments with code?

Patater commented on api/inc/register_gateway.h:
```diff
@@ -17,104 +17,208 @@
+ * (or their respective defaults if they have not been provided).
+ */
+#define uvisor_read(box_name, addr, op, msk) \
+    ({ \
```
Do all our compilers support the `({` syntax okay?

AlessandroA replied:
Yup. ARMCC is not officially supported yet by uVisor but it has the same extension (it's a GNU extension).

Patater replied:
So, we aren't using `offsetof` because of ARMCC. But we are using `({`, even though ARMCC, because ARMCC is not supported. Both these uses are in api/inc headers.

AlessandroA replied:
What I meant is that ARMCC supports the GNU extensions, hence it supports `({ ... })`. It does not include the GCC intrinsics/builtins though, so it does not have `offsetof`.

Patater commented on api/inc/register_gateway.h:
```diff
@@ -17,104 +17,208 @@
+            .mask       = msk, \
+            .operation  = op, \
+            .bxlr       = BXLR  \
+        }; \
```
Note that you didn't need to add `__attribute__((aligned(2))` here, because you are doing `| 1` and the object might (slight chance) not be aligned to a 2 byte boundary in which case no change to the address of the object needs to change for it to be a valid thumb-2 target.

AlessandroA replied:
Sorry I don't get the comment!

AlessandroA replied:
Ack, now I get it. Yes, initially I was thinking about the old 4-byte alignment requirements and added an attribute to enforce it, but this is not required for the reason you mentioned.

niklas-arm replied:
Hm, the ARMv7-M ARM states "All instruction fetches must be halfword-aligned." (page 48). We are not returning from an exception here therefore:
> - bit [0] of the value is ignored when loading the PC using an ADD or MOV instruction
> - a BLX, BX, LDR to the PC, POP or LDM including the PC instruction will cause an INVSTATE UsageFault if bit [0] of the value loaded is zero
> - loading the PC with a value from a memory location whose address is not word aligned is UNPREDICTABLE.

Since we cannot be sure of the actual mechanism the compiler uses to execute that struct (probably `blx` though), I'd actually make it 4-byte aligned.

Patater replied:
Actually, yes. It is required still. Ignore my nonsense. The CPU would read from a memory address 0x1234566 for a jump target of 0x1234567. This obviously won't work right if the CPU instruction actually starts at 0x1234567. I'd recommend you add an explicit alignment on the `TRegisterGateway`, `__attribute__((aligned(4))`.

AlessandroA replied:
Thanks, you're right @niklas-arm and @Patater ! Fixed.

Patater commented on api/inc/register_gateway_exports.h:
```diff
@@ -0,0 +1,57 @@
+    uint32_t mask;
+    uint16_t operation;
+    uint16_t bxlr;
+} UVISOR_PACKED TRegisterGateway;
```
Add a comment explaining why this must be packed, including the reason that there must be no padding before `svc_opcode`.

AlessandroA replied:
Done.

Patater replied:
Force an alignment here. This struct must always be located at an executable location, which means `__attribute__((aligned(4))`.

AlessandroA replied:
Done.

AlessandroA replied:
Actually, I put it into the typedef as you suggested: It's a requirement of the gateway to be always aligned.

AlessandroA commented:
* Addressed second round of review by @Patater

AlessandroA commented:
* Addressed third round of review by @Patater and @niklas-arm.

AlessandroA commented:
* Rebased after #227 has been merged.

Note: The commit list is shown out-of-order by GH, but the actual branch commit order is correct.

meriac commented:
@AlessandroA @niklas-arm @Patater : page_allocator.c could have a conditional .h include depending on whether the .c is compiled in a uvisor context or in unsupported. The unsupported header file does not need to be included in the supported tree and vice versa. Potentially the  header file name can be identical for bot supported and unsupported ("page_allocator_config.h") and have different content for supported/unsupported.

meriac commented on api/inc/register_gateway.h:
```diff
@@ -17,104 +17,210 @@
+/** Get the offset of a struct member.
+ * @internal
+ */
+#define OFFSETOF(type, member) ((uint32_t) (&(((type *)(0))->member)))
```
@AlessandroA : OFFSET_OF instead ?

niklas-arm replied:
You may have to remove this [`#define OFFSETOF`](https://github.com/ARMmbed/uvisor/blob/dev/api/rtx/src/secure_allocator.c#L33).

AlessandroA replied:
I used `OFFSETOF` because it's the macro implementation of the GCC built-in `offsetof`. Do you prefer the other name? It's only used here anyway.

meriac replied:
@AlessandroA we needs to namespace it anyway for uVisor, right - UVISOR_OFFSET_OF ?

meriac commented on api/inc/register_gateway.h:
```diff
@@ -17,104 +17,210 @@
+ * @returns the 16-bit encoding of the B.N <label> instruction.
+ */
+#define BRANCH_OPCODE(instr, label) \
+    (uint16_t) (0xE000 | (uint8_t) ((((uint32_t) (label) - (uint32_t) (instr + 4)) / 2) & 0xFF))
```
@AlessandroA : Shouldn't that be "(uint32_t) instr + 4" ?

AlessandroA replied:
Good point! Fixed.

AlessandroA commented:
* Added the issue #241 to track the runtime OR operation on the gateway function pointer.

meriac commented on core/system/src/register_gateway.c:
```diff
@@ -0,0 +1,125 @@
+     * register gateway pointers section. */
+    uint32_t * ptr = __uvisor_config.register_gateway_ptr_start;
+    bool found = false;
+    for (; ptr < __uvisor_config.register_gateway_ptr_end; ptr++) {
```
@AlessandroA : I hear unicorns crying... what about the both-way links ?

AlessandroA replied:
Yeah I know. The circular links didn't work unfortunately.

meriac replied:
How about we remove it then and rethink it ?

meriac commented on core/system/src/register_gateway.c:
```diff
@@ -0,0 +1,125 @@
+            HALT_ERROR(SANITY_CHECK_FAILED, "Register gateway 0x%08X is not in public flash.",
+                       (uint32_t) register_gateway);
+        }
+    }
```
@AlessandroA: too heavy-handed - please just verify the alignment of the start and that the start is a vmpu_public_flash_addr

meriac commented on core/system/src/register_gateway.c:
```diff
@@ -0,0 +1,127 @@
+     *   0x42000000 - 0x43FFFFFF: Peripheral alias
+     *   0xE00FF000 - 0xE00FFFFF: Custom ROM Table */
+    uint32_t address = register_gateway->address;
+    if (((address & VMPU_PERIPH_MASK) != VMPU_PERIPH_START) &&
```
@Alessandro you can combine both into one rule -and ignore the 1MB hole inbetween.

AlessandroA replied:
Done.

meriac commented on core/system/src/register_gateway.c:
```diff
@@ -0,0 +1,127 @@
+ *                              flash.
+ * @returns the extracted address of the register gateway.
+ */
+static TRegisterGateway * register_gateway_check(TRegisterGateway * register_gateway)
```
@AlessandroA Please move the HALT_ERRORS out of this function, create an approriate error.

AlessandroA replied:
Done.

meriac commented on core/system/src/register_gateway.c:
```diff
@@ -0,0 +1,127 @@
+
+    /* If none of the above checks failed, the same pointer to the register
+     * gateway is returned to the caller. */
+    return register_gateway;
```
just return the error code - it's confusing to the parameter. Please make the input parameters const const.

AlessandroA replied:
Done.

meriac commented on core/system/src/register_gateway.c:
```diff
@@ -0,0 +1,127 @@
+     * destination address from it. */
+    /* This function halts if it finds an error. */
+    TRegisterGateway * register_gateway = register_gateway_check((TRegisterGateway *) svc_pc);
+
```
Do the halt_error here including the result - access denied is fine.

AlessandroA replied:
Done.

AlessandroA commented:
* Addressed latest comments from @meriac. Now `register_gateway_check()` returns a status code. It's the caller that halts the system with a `NOT_ALLOWED` error.

meriac commented on core/system/inc/mpu/vmpu.h:
```diff
@@ -101,8 +101,17 @@ static UVISOR_FORCEINLINE int vmpu_sram_addr(uint32_t addr)
+#define VMPU_ROMTABLE_START 0xE00FF000UL
+#define VMPU_ROMTABLE_END   0xE00FFFFFUL
+#define VMPU_ROMTABLE_MASK  0xFFFFF000UL
 
```
@AlessandroA remove unused  VMPU_PERIPH_END and VMPU_PERIPH_BITBAND_END and VMPU_ROMTABLE_END symbols ? Create the wrong impression that the end can be non-aligned.
